src/Bounds.cpp

```diff
@@ -856,8 +856,27 @@ class Bounds : public IRVisitor {
             assert(op->args.size() == 3);
             op->args[1].accept(this);
         } else if (op->is_intrinsic(Call::shift_left) ||
-                   op->is_intrinsic(Call::shift_right) ||
-                   op->is_intrinsic(Call::bitwise_and)) {
+                   op->is_intrinsic(Call::shift_right)) {
+            assert(op->args.size() == 2);
+            Expr simplified = simplify(op);
+            if (!equal(simplified, op)) {
+                simplified.accept(this);
+            } else {
+                op->args[0].accept(this);
+                Interval a = interval;
+                op->args[1].accept(this);
+                Interval b = interval;
+                if (a.is_bounded() && b.is_bounded()) {
+                    Expr shift_min = op->is_intrinsic(Call::shift_left) ? b.min : b.max;
+                    Expr shift_max = op->is_intrinsic(Call::shift_left) ? b.max : b.min;
+                    interval.min = Call::make(t, op->name, {a.min, shift_min}, Call::PureIntrinsic);
```

Comment:
I don't think this logic is correct. Shifting left is frequently used to drop high bits, which makes it non-monotonic. A lot of the rest of the bounds visitor is careful to prove overflow cannot occur before doing things like this. An easy fix would be to add ```op->type.is_int() && op->type.bits() >= 32``` to the condition. Those are the types for which overflow is UB.

Reply:
Thanks Andrew - yes, you are completely correct. I detected this while updating the test and am now writing overflow detection logic; my plan had been to insert expressions to be evaluated at runtime that determine if the given expression would overflow, and instead return the bounds of the type when that would occur. I believe that is the most correctness-friendly thing we can do in this circumstance.

That work is taking a bit of time, because there are a variety of circumstances which need to be handled slightly differently; in particular, Halide permits negative shift values, so I need to deal with positive left shifts and negative right shifts, on both unsigned and signed values. I'll update the PR when it's done.

Reply:
Quick question - I *thought* that Halide permits signed shift values (in that the language does not assert or seem to forbid it in documentation), but I'm now unsure. I have code that handles all other cases correctly, but I see that when I perform `-1234 >> -2`, this produces -1, not -4936. I see in the STMT I am generating that my code simplifies down to:
`let input.min.0.required.s = shift_right(-1234, -2)`. I also see that in LLVM IR, all of `shl`, `lshr`, and `ashr` assume that the second operand is unsigned (and I do see that an `ashr` is being generated to perform this shift).

I hypothesize that shifting by negative values is in fact UB in Halide, and that LLVM has interpreted my `-2` as `0xfffe` (or `65534`), and is shifting off all the real bits of my value and filling with the sign bit, resulting in `-1`. Thoughts?